src/VisualStudio/Core/Def/Implementation/EditAndContinue/VisualStudioDebugStateChangeListener.cs

```diff
@@ -11,78 +11,188 @@
+#pragma warning restore
+
         /// <summary>
         /// Called by the debugger when a debugging session starts and managed debugging is being used.
         /// </summary>
-        public void StartDebugging(DebugSessionOptions options)
+        public async Task StartDebuggingAsync(Dbg.DebugSessionOptions options, CancellationToken cancellationToken)
         {
             _debuggingService.OnBeforeDebuggingStateChanged(DebuggingState.Design, DebuggingState.Run);
+            _disabled = (options & Dbg.DebugSessionOptions.EditAndContinueDisabled) != 0;
+
+            if (_disabled)
+            {
+                return;
+            }
 
-            if ((options & DebugSessionOptions.EditAndContinueDisabled) == 0)
+            try
             {
-                _encService = _workspace.Services.GetRequiredService<IEditAndContinueWorkspaceService>();
-                _encService.StartDebuggingSession(_workspace.CurrentSolution);
+                var solution = _proxy.Workspace.CurrentSolution;
+                await _proxy.StartDebuggingSessionAsync(solution, cancellationToken).ConfigureAwait(false);
             }
-            else
+            catch (Exception e) when (FatalError.ReportWithoutCrashUnlessCanceled(e))
             {
-                _encService = null;
+                _disabled = true;
             }
         }
 
-        public void EnterBreakState(IManagedActiveStatementProvider activeStatementProvider)
+        public async Task EnterBreakStateAsync(Dbg.IManagedActiveStatementProvider activeStatementProvider, CancellationToken cancellationToken)
         {
             _debuggingService.OnBeforeDebuggingStateChanged(DebuggingState.Run, DebuggingState.Break);
 
-            _encService?.StartEditSession(async cancellationToken =>
+            if (_disabled)
             {
-                var infos = await activeStatementProvider.GetActiveStatementsAsync(cancellationToken).ConfigureAwait(false);
-                return infos.SelectAsArray(ModuleUtilities.ToActiveStatementDebugInfo);
-            });
+                return;
+            }
+
+            try
+            {
+                _editSessionConnection = await _proxy.StartEditSessionAsync(
+                    _diagnosticService,
+                    activeStatementProvider: async cancellationToken =>
+                    {
+                        var infos = await activeStatementProvider.GetActiveStatementsAsync(cancellationToken).ConfigureAwait(false);
+                        return infos.SelectAsArray(ModuleUtilities.ToActiveStatementDebugInfo);
+                    },
+                    debuggeeModuleMetadataProvider: _managedModuleInfoProvider,
+                    cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception e) when (FatalError.ReportWithoutCrashUnlessCanceled(e))
+            {
+                _disabled = true;
+            }
```

Comment:
as an aside, a lot of these catches bother me in the sense of we NFW, but then the user may be in some (non-broken) state, but unclear on why something they want to be working isn't.  can we file issue to at least yellow-bar or something to let them know that something went wrong, but that otherwise they can continue?

Reply:
I'll address it when we move this to ISB.